networktable.c

```diff
@@ -0,0 +1,337 @@
+
+		// Get ID of this device in our network database. If it cannot be found, then this is a new device
+		// We match both IP *and* MAC address
+		// Same MAC, two IPs: Non-deterministic DHCP server, treat as two entries
+		// Same IP, two MACs: Either non-deterministic DHCP server or (almost) full DHCP address pool
+		// We can run this SELECT inside the currently active transaction as only the
+		// changed to the database are collected for latter commitment. Read-only access
+		// such as this SELECT command will be executed immediately on the database.
+		char* querystr = NULL;
+		int ret = asprintf(&querystr, "SELECT id FROM network WHERE ip = \"%s\" AND hwaddr = \"%s\";", ip, hwaddr);
+		if(querystr == NULL || ret < 0)
+		{
+			logg("Memory allocation failed in parse_arp_cache (%i)", ret);
+			break;
+		}
+
+		// Perform SQL query
+		int dbID = db_query_int(querystr);
+		free(querystr);
+
+		if(dbID == -2)
+		{
+			// SQLite error
+			break;
+		}
+
+		// If we reach this point, we can check if this client
+		// is known to pihole-FTL
+		// false = do not create a new record if the client is
+		//         unknown (only DNS requesting clients do this)
+		lock_shm();
+		int clientID = findClientID(ip, false);
+		unlock_shm();
+
+		// This client is known (by its IP address) to pihole-FTL if
+		// findClientID() returned a non-negative index
+		bool clientKnown = clientID >= 0;
+
+		// Get hostname of this client if the client is known
+		char *hostname = NULL;
+		if(clientKnown)
+		{
+			validate_access("clients", clientID, true, __LINE__, __FUNCTION__, __FILE__);
+			hostname = getstr(clients[clientID].namepos);
+		}
+
+		// Device not in database, add new entry
+		if(dbID == -1)
+		{
+			char* macVendor = getMACVendor(hwaddr);
+			dbquery("INSERT INTO network "\
+			        "(ip,hwaddr,interface,firstSeen,lastQuery,numQueries,name,macVendor) "\
+			        "VALUES (\"%s\",\"%s\",\"%s\",%lu, %ld, %u, \"%s\", \"%s\");",\
+			        ip, hwaddr, iface, now,
+			        clientKnown ? clients[clientID].lastQuery : 0L,
+			        clientKnown ? clients[clientID].numQueriesARP : 0u,
+			        hostname == NULL ? "" : hostname,
+			        macVendor);
+			if(strlen(macVendor) > 0)
+				free(macVendor);
+		}
+		// Device in database AND client known to Pi-hole
+		else if(clientKnown)
+		{
+			// Update lastQuery. Only use new value if larger
+			// clients[clientID].lastQuery may be zero if this
+			// client is only known from a database entry but has
+			// not been seen since then
+			dbquery("UPDATE network "\
+			        "SET lastQuery = MAX(lastQuery, %ld) "\
+			        "WHERE id = %i;",\
+			        clients[clientID].lastQuery, dbID);
+
+			// Update numQueries. Add queries seen since last update
+			// and reset counter afterwards
+			dbquery("UPDATE network "\
+			        "SET numQueries = numQueries + %u "\
+			        "WHERE id = %i;",\
+			        clients[clientID].numQueriesARP, dbID);
+			clients[clientID].numQueriesARP = 0;
```

Comment:
Because the normal query count gets populated by the imported database queries. This means that when having 2000 queries in the history and restarting FTL 2 times, 4000 additional queries would be added to the total counter. This would obviously be incorrect. Hence this new counter which is only incremented on (new!) incoming queries and not on database imported ones.

Reply:
The query count could be just the client total count, instead of incrementing. In your scenario, it would still show up as 2000 queries because that is how many queries the client has. This would mean the `numQueries` column would go away from the table.

Reply:
Just to get this right: You mean something like `SELECT count(*) FROM queries WHERE ip = "whatever";`? I think this may be very slow for very large databases. Also, the current implementation is *both* IP *and* MAC address sensitive and queries made by a different device (that had the same IP at some time before) would still be separated.

Reply:
This would only be the momentary count (most recent 24 hours) whereas my idea was to have a global counter (queries *ever* seen from this device). The two numbers are conceptually different.

Reply:
Ah, that would make sense.